content/docs/iac/using-pulumi/extending-pulumi/build-a-provider.md

```diff
@@ -0,0 +1,756 @@
+- **Configure** – initializes provider-wide settings (e.g. credentials)
+
+### How a provider runs
+
+Pulumi providers are [gRPC](https://grpc.io/) servers that respond to commands from the Pulumi engine. When a Pulumi program runs (e.g. `pulumi up`), the engine connects to the provider process and sends instructions to create, update, or delete resources via calls to the provider interface implementation.
+
+### Configuration, secrets, outputs, and state
+
+Beyond the core functions involved with managing resources, there are a number of other aspects to a Pulumi provider. It is necessary to configure the provider, pass secrets, return output values, and store the state of resources. The Pulumi provider interface has built-in facilities for all of those concerns:
+
+- **Configuration and Secrets**: Set via [Pulumi ESC](/docs/esc/) [environments](/docs/esc/environments/) and/or `pulumi config`. Encrypted secrets and configuration values are passed to the provider at runtime.
+- **Outputs**: Providers return outputs from resources, which can be referenced by other resources.
+- **State**: Pulumi maintains resource state to track dependencies and detect changes.
+
+### Handling errors and failures
+
+Providers should report meaningful error messages. It’s important to handle transient failures and make operations [idempotent](https://en.wikipedia.org/wiki/Idempotence) to avoid inconsistent states.
+
+### The provider schema
+
+A provider schema defines the resources, their input and output properties, data sources, and configuration options. This schema enables Pulumi to generate SDKs for multiple languages and ensures consistency across them, as well as providing documentation.
+
+{{% notes type="info" %}}
+
+Historically it was necessary to hand-author and maintain the `schema.json` file that accompanied your provider implementation, however, now most of this is handled on-the-fly by the [Pulumi Provider SDK](/docs/iac/using-pulumi/extending-pulumi/pulumi-provider-sdk/) and the file is no longer necessary.
+
+{{% /notes %}}
+
+## Language support and the Pulumi Provider SDK
+
+Pulumi providers can be written in Go, TypeScript, .NET, and Java, and used in any Pulumi program, in any supported language. The [Pulumi Provider SDK](/docs/iac/using-pulumi/extending-pulumi/pulumi-provider-sdk/) is a framework for building providers in Go. We strongly recommend using the SDK, as it is the most full-featured and streamlined way to create a new provider.
+
+Some advantages of using the Pulumi Provider SDK:
+
+- **Minimal Code Required**: You define your resource types and implementation using Go structs and methods, and the SDK handles the rest (RPC, auto-generated schema for multi-language support, etc).
+- **Includes a Testing Framework**: Testing custom providers is made much easier with the SDK's built-in testing framework.
+- **Middleware Support**: Enhances providers with layers like token dispatch, schema generation, and cancellation propagation.
```

Comment:
"token dispatch" is a phrase seen in the code too, and I find it a bit confusing. It means "type tokens" because the dispatch logic is keyed off the type token.  The phrase "dispatch logic" here would make better sense.

Reply:
thank you! tbh i was also a bit confused by that terminology.